Fix Connection Pool

caleblloyd opened the pull request:
I've taken a stab at fixing the connection pool.  I tested this implementation with MySqlConnector.Performance, but it seems to hang the SideBySide.New tests.  SideBySide.New does not print which test is currently running, so I'm not sure how to debug.

@bgrainger can you take a look at the SideBySide.New tests and see if you can tell why they are hanging?  If you can add logging to it to print the method it's in, I can resolve the bug.  I tried briefly but I'm not very familiar with XUnit.

Async targets are running stable at 1500 Requests Per Second in the Performance Tests.  Very nice!!!  I'm running on an Intel Core i5-6260U, load test, web app, and database all running on the same machine.  Should do better on a dedicated app server.

	Rate = 1500 RPS
	Duration = 10s
	Targets = targets-async.txt

	Requests      [total, rate]            15000, 1500.10
	Duration      [total, attack, wait]    10.000748548s, 9.999322811s, 1.425737ms
	Latencies     [mean, 50, 95, 99, max]  23.299198ms, 2.395148ms, 168.782969ms, 271.810346ms, 395.213902ms
	Bytes In      [total, mean]            173108268, 11540.55
	Bytes Out     [total, mean]            5561250, 370.75
	Success       [ratio]                  100.00%
	Status Codes  [code:count]             200:15000

bgrainger commented:
I can look at this soon. Can you summarise the overall intent of these changes (so I can understand the purpose and review them according to that)?

You say this is "fixing the connection pool", but I'm not actually sure what is broken with it (#38 didn't provide any details). Is it something that can be simulated and verified in an integration test?

bgrainger commented on src/MySqlConnector/MySqlClient/ConnectionPool.cs:
```diff
@@ -7,101 +9,191 @@ namespace MySql.Data.MySqlClient
+				m_session_semaphore.Release(m_maximumSize);
+
+				// re-open the minimum sessions
+				await OpenMinimumSessions();
```
I'm not sure that `ClearAsync` should reopen connections; AFAIK it's commonly used to make sure all connections *are* shut down e.g., when exiting a program.

caleblloyd replied:
Good point, this should be taken out

bgrainger commented on src/MySqlConnector/MySqlClient/ConnectionPool.cs:
```diff
@@ -7,101 +9,191 @@ namespace MySql.Data.MySqlClient
+
+			ConnectionPool pool;
+			// try without the lock
+			if (!s_pools.TryGetValue(key, out pool))
```
This isn't threadsafe.

caleblloyd replied:
This is a 2-part check. First part tries acquiring without a lock.  If it gets null, then it tries acquiring with lock.  Thought behind this is GetPoolAsync is called once per connection but it is only ever instantiated once in an application (and never removed).  This reduces lock contention.

bgrainger replied:
`s_pools.Add` could be called (on line 126) from a different thread while this thread calls `TryGetValue`. That is not threadsafe and will crash if the dictionary's internal state is inconsistent.

caleblloyd replied:
Ah.  Would you prefer that I change it to a ConcurrentDictionary or just use the Async lock every time?

caleblloyd replied:
Fixed by switching to ConcurrentDictionary

bgrainger replied:
Using a well-tested framework class is a good approach. Can try to replace it if performance profiling indicates it's a bottleneck.

bgrainger commented on src/MySqlConnector/MySqlClient/ConnectionPool.cs:
```diff
@@ -7,101 +9,191 @@ namespace MySql.Data.MySqlClient
+				return session;
 			}
-			finally
+			catch
```
Why not keep it as `finally`?

caleblloyd replied:
The `m_session_semaphore` is now set to the size of `m_max_connections`.  It is held until the connection is returned to the pool in the Return(session) method.  If there is an exception, it should be returned to the pool immediately since the session never leaves the pool.

caleblloyd commented:
> Can you summarise the overall intent of these changes (so I can understand the purpose and review them according to that)?

The current implementation does not track the number of Connections that are leased out of the pool.  Instead, it just checks to see if it over-created sessions once they are returned.  If load is high and sessions aren't getting returned to the pool, the current implementation will hand out unlimited connections and overwhelm the database.

This PR changes `m_session_semaphore` to track how many Connections are leased out of the pool.  It counts them back into the pool in Return(session).

I moved a few methods from MySqlConnection to MySqlSession because the ConnectionPool needed to access them, and they seemed more appropriate in MySqlSession.

caleblloyd commented:
Added the fix for making s_pools a ConcurrentDictionary and not re-opening minimum connections on ClearPools

Still having issues with SideBySide.New hanging though.  Any luck looking at that?

bgrainger commented:
> If load is high and sessions aren't getting returned to the pool fast enough, the current implementation will hand out unlimited connections and overwhelm the database.

That makes total sense. I need to find out why MySql.Data does in this instance... just block until a connection is available? Or fail immediately with an exception?

caleblloyd commented:
It definitely blocks until a connection is available. I confirmed by load testing it.

bgrainger commented:
Although it will cause merge conflicts, I'm going to start breaking this up into smaller commits, to narrow down the cause of test failures.

First pushed: moving `ConnectAsync` etc. into `MySqlSession`: 54e3a8df04af2030d990305ce96a4d88aeb8cd7c

bgrainger commented:
I abandoned the `Semaphore` approach and went with a simple `Monitor`/`lock` approach that let me more easily maintain invariants: 2dd346cee28c95e30b3977264ea2b10e72f4ece6

(The issue I ran into with the semaphore is that I couldn't atomically wait for it and increment a counter, so I couldn't find a way to assert invariants about the state of the pool--i.e., the number of active+idle connections--to verify it was behaving correctly. The downside with the new approach is that a thread will actually block in `GetSessionAsync` while it waits for a connection to become available. However, I'd like to make it correct first, then optimise it to be non-blocking later on.)

It doesn't implement `MinimumPoolSize` yet, but that can come later.

bgrainger commented:
The new code can maintain 400 req/s:

```
PS> .\stress.ps1 400 20s async
Rate = 400 RPS
Duration = 20s
Targets = targets-async.txt
Requests      [total, rate]            8000, 400.05
Duration      [total, attack, wait]    20.0086398s, 19.9975s, 11.1398ms
Latencies     [mean, 50, 95, 99, max]  5.626899ms, 5.1117ms, 11.6336ms, 14.1336ms, 19.6386ms
Bytes In      [total, mean]            92626005, 11578.25
Bytes Out     [total, mean]            2978000, 372.25
Success       [ratio]                  100.00%
Status Codes  [code:count]             200:8000
Error Set:
```

I am getting some failures above that, so will dig in further.

caleblloyd commented:
@bgrainger I'm seeing the same thing on perf tests.

Spinlocks are pretty detrimental to performance when you get a lot of threads waiting at load.  I think we should try to switch it back to SemaphoreSlim if possible

bgrainger commented:
I'm going to try porting the new code to `SemaphoreSlim` (if I can determine a good analogue to `Monitor.Pulse`).

bgrainger commented:
Switched to SemaphoreSlim: 187eb0bdb27010b38c780508b0575e2a3b5f332d

I have an idea for implementing an alternative to `Monitor.Pulse`, but don't have the time to implement it right now.

It still fails at 500 req/s but it doesn't grow to 100s or 1000s of blocked threads (as it did with the previous implementation).

greghroberts commented:
Npgsql's implementation is certainly something to refer to..
https://github.com/npgsql/npgsql/blob/74651fc696dbb84b46705de896870c038d4fd5b0/src/Npgsql/PoolManager.cs

Also System.Data.SqlClient 
https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnectionFactory.cs

caleblloyd commented:
Is there a reason we need to be able to keep track of what state the connections are in inside of the pool?  What's wrong with counting them out of the pool and counting them back into the pool with a `SemaphoreSlim` with max size of max_connections?

I've also corrected the behavior of `Pooling=False`.  There should be no connection pool at all with `Pooling=False`, the app should manage it's own connections at this point.

I've updated this PR to be based off the `tests` branch starting at 187eb0bdb27010b38c780508b0575e2a3b5f332d but reverted it to use the count in / count out method.  I'm maintaining 1500 RPS easily with this approach.

Still having an issue with some of the SideBySide.New tests.  The tests don't seem to be running serially and they don't print the test name, so it's very hard to track down deadlocks in the tests.  Any way we can at least get the tests to log their name?

caleblloyd commented on src/MySqlConnector/Serialization/MySqlSession.cs:
```diff
@@ -164,41 +170,44 @@ private void VerifyConnected()
 					{
-						socket = new Socket(ipAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
-						using (cancellationToken.Register(() => socket.Dispose()))
+						var connectTasks = new List<Task>
```
The Socket tasks do not take a cancellation token as an argument, so the Timeout Cancellation token will not do anything.  In order to find a timeout, I have leveraged Task.WaitAny with a timeout task, and checked the socket status.

bgrainger replied:
What was wrong with registering a handle on the cancellationToken that called `Socket.Dispose`?

As per [MSDN](https://msdn.microsoft.com/en-us/library/tad07yt6.aspx):

> To cancel a pending call to the BeginConnect method, close the Socket. When the Close method is called while an asynchronous operation is in progress, the callback provided to the BeginConnect method is called. A subsequent call to the EndConnect method will throw an ObjectDisposedException to indicate that the operation has been cancelled. 

(The Socket async API has lots of "oddities" and this is one of them.)

bgrainger replied:
(.NET Core doesn't have `Close`, but it does have `Dispose` which I checked does the same thing.)

caleblloyd replied:
I can't remember exactly - I'd have to go back and re-test.  Does an `await` command called on `socket.OpenAsync` without a cancellationToken block and not consider the handle on the CancellationToken?

caleblloyd commented on tests/SideBySide.New/ConnectionPool.cs:
```diff
@@ -80,37 +80,7 @@ public void ResetConnection(object connectionReset, int poolSize, long expected)
 		}
 
 		[Fact]
-		public async Task ExhaustConnectionPoolWithTimeout()
```
I believe this test to be incorrect.  I think that ConnectionTimeout is a timeout for a TCP Connection to be made to the MySQL server.  I don't think it applies to a pooled connection being retrieved from the pool.

bgrainger replied:
MySql.Data uses this value to time out the connection when the pool is full. (This test passes when compiled with mysql-connector-net.) This behaviour makes sense to me and I think we should clone it. (The alternatives would be failing immediately or blocking forever, and none of those seems preferable.)

caleblloyd replied:
Ok, we should reimplement this using a timeout CancellationToken inside of `ConnectionPool.GetSessionAsync` then

caleblloyd commented on tests/SideBySide.New/ConnectionPool.cs:
```diff
@@ -148,9 +118,5 @@ public void ResetConnection(object connectionReset, int poolSize, long expected)
 
-		public void Dispose()
-		{
-			MySqlConnection.ClearAllPools();
```
XUnit instantiates all test classes at the very beginning.  Since pools are shared among tests, `ClearAllPools` wreaks havoc on other tests that are still running.

bgrainger replied:
I agree that clearing pools at the end of each test isn't ideal, but I'm not clear on how this "wreaks havoc"; are there tests that specifically rely on idle connections being available? 

This may actually indicate a bug in `ClearAllPools`; it shouldn't be implemented as a "destructive" operation for existing open connections but as per the [documentation](https://dev.mysql.com/doc/dev/connector-net/html/M_MySql_Data_MySqlClient_MySqlConnection_ClearPool.htm)

> ClearPool clears the connection pool that is associated with the connection. If additional connections associated with connection are in use at the time of the call, they are marked appropriately and are discarded (instead of being returned to the pool) when Close is called on them.

(I don't think my implementation guarantees that, so I'll need to fix that up.)

caleblloyd replied:
Ah, good find!  This should be an easy fix then - let's open another issue for it after this is merged.

caleblloyd commented on tests/SideBySide.New/DatabaseFixture.cs:
```diff
@@ -22,7 +22,6 @@ protected virtual void Dispose(bool disposing)
 			if (disposing)
 			{
 				Connection.Dispose();
-				MySqlConnection.ClearAllPools();
```
XUnit instantiates all test classes at the very beginning.  Since pools are shared among tests, `ClearAllPools` wreaks havoc on other tests that are still running.

caleblloyd commented:
I've gotten tests working locally and on appveyor with this implementation.  Not sure why Travis is failing.  Running solid at 1500 RPS.  @bgrainger please read my code review comments.

bgrainger commented:
> Not sure why Travis is failing.

Very odd; can't tell yet if something changed in the environment that's causing that failure.

bgrainger commented:
> Running solid at 1500 RPS.

Awesome! :+1: Will merge this once I can get the tests figured out.

bgrainger commented:
Shipped this in [0.1.0-alpha21](https://www.nuget.org/packages/MySqlConnector/0.1.0-alpha21).
